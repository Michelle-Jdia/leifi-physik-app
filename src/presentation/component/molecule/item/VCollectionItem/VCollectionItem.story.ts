import type { Meta, StoryFn } from '@storybook/vue3';
import { collection } from '@/presentation/static/collection';
import VCollectionItem from '@/molecule/item/VCollectionItem/VCollectionItem.vue';

export default {
    component: VCollectionItem,
} as Meta<typeof VCollectionItem>;

const Template: StoryFn<typeof VCollectionItem> = (args) => ({
    components: {
        VCollectionItem,
    },
    setup() {
        return {
            args,
        };
    },
    template: `
        <v-collection-item v-bind="args"></v-collection-item>
    `,
});

export const Base = Template.bind({});

export const WithIcon = Template.bind({});

export const WithNotification = Template.bind({});

export const WithLearnStatus = Template.bind({});

Base.args = {
    privateTitle: collection.private_title,
    subtitle: collection.title,
    isSubscribed: false,
};

WithIcon.args = {
    privateTitle: collection.private_title,
    subtitle: collection.title,
    isSubscribed: true,
};

WithNotification.args = {
    privateTitle: collection.private_title,
    subtitle: collection.title,
    isSubscribed: true,
    withNotification: true,
};

WithLearnStatus.args = {
    privateTitle: collection.private_title,
    subtitle: collection.title,
    isLearnMode: true,
    learnStatus: '3/3 Inhalten gelernt',
};
