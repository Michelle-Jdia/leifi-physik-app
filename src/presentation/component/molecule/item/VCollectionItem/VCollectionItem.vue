<script setup lang="ts">
import type { ClassValue } from 'clsx';
import { cn } from '@/presentation/helper/style';
import VBadge from '@/atom/badge/VBadge/VBadge.vue';
import VIcon from '@/atom/icon/VIcon/VIcon.vue';

interface Props {
    subtitle: string;
    privateTitle: string;
    isSubscribed?: boolean;
    isLearnMode?: boolean;
    learnStatus?: string;
    withNotification?: boolean;
    modifier?: ClassValue;
}

defineProps<Props>();
</script>

<template>
    <div
        :class="
            cn(
                'collection-item v-relative v-flex v-items-start v-justify-between v-rounded-sm v-border v-border-gray-20 v-bg-white v-p-box v-shadow',
                modifier,
            )
        "
    >
        <div>
            <h3 class="v-h3">{{ privateTitle }}</h3>

            <p class="v-p v-mt-1 v-text-gray">{{ subtitle }}</p>
        </div>

        <v-icon
            v-if="isSubscribed && withNotification"
            size="sm"
            data-testid="collection-item-notification"
            src="icon/ellipse.svg"
            modifier="v-absolute v-right-0 v-top-0 v--translate-y-2/4 v-translate-x-2/4"
        />

        <v-icon
            v-if="isSubscribed"
            size="sm"
            data-testid="collection-item-feed"
            src="icon/bond.svg"
            with-space-left
            modifier="v-text-primary"
        />

        <v-badge v-if="isLearnMode && learnStatus" modifier="v-ml-box-sm">
            {{ learnStatus }}
        </v-badge>
    </div>
</template>
