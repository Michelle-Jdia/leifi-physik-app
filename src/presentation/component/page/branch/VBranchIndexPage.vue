<script setup lang="ts">
import type { SliderItem } from '@/molecule/slider/VTopicBoxSlider/VTopicBoxSlider.vue';
import type { RefresherCustomEvent } from '@ionic/vue';
import type { ComputedRef } from 'vue';
import { onIonViewWillEnter } from '@ionic/vue';
import { computed } from 'vue';
import { createRefresherHandler } from '@/connection/helper/refresher';
import { RouteName } from '@/connection/router/routeName';
import { useBranchState } from '@/state/branchState';
import { navigateToSearch } from '@/store/searchStore';
import { createTopicSlider } from '@/presentation/helper/topic';
import VBranchHeader from '@/layout/header/VBranchHeader/VBranchHeader.vue';
import VPage from '@/layout/page/VPage.vue';
import VGlobalSearchInput from '@/container/form/VGlobalSearchInput/VGlobalSearchInput.vue';
import VTopicBoxSlider from '@/molecule/slider/VTopicBoxSlider/VTopicBoxSlider.vue';
import VBranchBox from '@/atom/box/VBranchBox/VBranchBox.vue';
import VLoader from '@/atom/loader/VLoader/VLoader.vue';
import VRefresher from '@/atom/refresher/VRefresher/VRefresher.vue';

type TopicSliderItem = Record<string, SliderItem[]>;

const branchState = useBranchState();

const sliders: ComputedRef<TopicSliderItem | void> = computed(() => {
    if (!branchState.topics) {
        return;
    }

    return branchState.topics.reduce((sliders: TopicSliderItem, topic) => {
        if (topic && topic[0]) {
            const branchId = topic[0].referenced_branch;

            if (!sliders[branchId]) {
                sliders[branchId] = topic.map(createTopicSlider);
            }
        }

        return sliders;
    }, {});
});

async function handleRefresh(event: RefresherCustomEvent): Promise<void> {
    await createRefresherHandler({
        event: event,
        callback: async () => {
            await branchState.getLiveTopicsAndBranches();
            await branchState.getTopics();
        },
    });
}

onIonViewWillEnter(branchState.getTopics);
</script>

<template>
    <v-page>
        <v-loader
            :logs="['useBranches', 'getTopicsByBranch', 'useBranchesLive', 'getTopicsByBranchLive']"
        />

        <template #header>
            <v-branch-header @on-submit="navigateToSearch">Alle Inhalte</v-branch-header>
        </template>

        <template #refresher>
            <v-refresher @ion-refresh="handleRefresh" />
        </template>

        <v-global-search-input modifier="v-mb-box-xl md:v-hidden" />

        <h1 class="v-h2 v-font-bold">Teilgebiete der Physik</h1>

        <div class="v-mt-box-xl v-grid v-grid-cols-2 v-gap-box-md md:v-grid-cols-4">
            <template v-for="branch in branchState.branches" :key="branch.name">
                <router-link :to="{ name: RouteName.BRANCH_ID, params: { id: branch.id } }">
                    <v-branch-box
                        :img="{
                            src: branch.icon?.url || '',
                            alt: branch.icon?.alt || '',
                        }"
                    >
                        {{ branch.name }}
                    </v-branch-box>
                </router-link>
            </template>
        </div>

        <v-topic-box-slider
            v-for="branch in branchState.branches"
            :key="branch.id"
            :color="branch.color"
            modifier="v-mt-box-md"
            :branch-img="{
                src: branch.icon?.url || '',
                alt: branch.icon?.alt || '',
            }"
            :link="{ name: RouteName.BRANCH_ID, params: { id: branch.id } }"
            :slider-items="(sliders && sliders[branch.id]) || []"
        >
            {{ branch.name }}
        </v-topic-box-slider>
    </v-page>
</template>
