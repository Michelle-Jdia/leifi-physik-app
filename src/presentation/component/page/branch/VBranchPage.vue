<script setup lang="ts">
import type { SliderItem } from '@/molecule/slider/VTopicBoxSlider/VTopicBoxSlider.vue';
import type { RefresherCustomEvent } from '@ionic/vue';
import type { ComputedRef } from 'vue';
import { computed } from 'vue';
import { createRefresherHandler } from '@/connection/helper/refresher';
import { useBranch } from '@/store/branchStore';
import { navigateToSearch } from '@/store/searchStore';
import {
    getTopicsByBranchLive,
    isTopicFirstDegree,
    isTopicSecondDegree,
    isTopicThirdDegree,
    TopicDegree,
    useTopicsByBranch,
} from '@/store/topicStore';
import { createTopicSlider } from '@/presentation/helper/topic';
import VBranchHeader from '@/layout/header/VBranchHeader/VBranchHeader.vue';
import VPage from '@/layout/page/VPage.vue';
import VTopicBoxSlider from '@/molecule/slider/VTopicBoxSlider/VTopicBoxSlider.vue';
import VLoader from '@/atom/loader/VLoader/VLoader.vue';
import VRefresher from '@/atom/refresher/VRefresher/VRefresher.vue';

interface Props {
    id: string;
}

const props = defineProps<Props>();

type TopicSliderItem = Record<string, SliderItem[]>;

const themeTitle = 'Themenbereiche';

const { data: branch, getData: getBranch } = useBranch({
    params: {
        id: props.id,
    },
});

const { data: topics, getData: getTopics } = useTopicsByBranch({
    params: {
        branchId: props.id,
    },
});

const { getData: getLiveTopics } = getTopicsByBranchLive({
    params: {
        branchId: props.id,
    },
});

const sliders: ComputedRef<TopicSliderItem | void> = computed(() => {
    if (!topics.value) {
        return;
    }

    const levelOneTopics = topics.value.filter(isTopicFirstDegree);
    const levelTwoTopics = topics.value.filter(isTopicSecondDegree);
    const levelThreeTopics = topics.value.filter(isTopicThirdDegree);

    return {
        levelOne: levelOneTopics.map(createTopicSlider),
        levelTwo: levelTwoTopics.map(createTopicSlider),
        levelThree: levelThreeTopics.map(createTopicSlider),
    };
});

function getTopicHeadline(key: string): TopicDegree | void {
    if (key === 'levelOne') {
        return TopicDegree.one;
    }

    if (key === 'levelTwo') {
        return TopicDegree.two;
    }

    if (key === 'levelThree') {
        return TopicDegree.three;
    }
}

async function handleRefresh(event: RefresherCustomEvent): Promise<void> {
    await createRefresherHandler({
        event: event,
        callback: async () => {
            await getLiveTopics();
            await getBranch();
            await getTopics();
        },
    });
}

getBranch();
getTopics();
</script>

<template>
    <v-page>
        <v-loader :logs="['useBranch', 'getTopicsByBranch']" />

        <template #header>
            <v-branch-header @on-submit="navigateToSearch">
                <template v-if="branch">
                    {{ branch.name }}
                </template>

                <template #sub-title>Teilgebiete</template>
            </v-branch-header>
        </template>

        <template #refresher>
            <v-refresher @ion-refresh="handleRefresh" />
        </template>

        <template v-if="sliders && branch">
            <template v-for="key in Object.keys(sliders)" :key="key">
                <div v-if="sliders[key] && sliders[key]?.length" class="v-mb-section">
                    <h2 class="v-h2">{{ getTopicHeadline(key) }}</h2>

                    <v-topic-box-slider
                        v-if="sliders[key]"
                        :key="branch.id"
                        modifier="v-mt-box-md"
                        :color="branch.color"
                        :branch-img="{
                            src: branch.icon?.url || '',
                            alt: branch.icon?.alt || '',
                        }"
                        :slider-items="sliders[key] || []"
                    >
                        {{ themeTitle }}
                    </v-topic-box-slider>
                </div>
            </template>
        </template>
    </v-page>
</template>
