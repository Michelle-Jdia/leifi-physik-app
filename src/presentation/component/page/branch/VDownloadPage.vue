<script setup lang="ts">
import type { Collection } from '@/data/type/app/collection';
import type { RefresherCustomEvent } from '@ionic/vue';
import type { Ref } from 'vue';
import { onIonViewWillEnter } from '@ionic/vue';
import { computed, ref, watch } from 'vue';
import { $t } from '@/application/translation';
import { ContentType } from '@/data/type/helper/contentType';
import { useDownloadTopicBranch } from '@/connection/helper/download';
import { createRefresherHandler } from '@/connection/helper/refresher';
import { RouteName } from '@/connection/router/routeName';
import { useArticleState } from '@/state/articleState';
import { findArticleInCollection } from '@/store/collectionStore';
import { useDownloadLive } from '@/store/downloadStore';
import { addDownloadToHistory } from '@/store/historyStore';
import { handleLinksClick } from '@/presentation/helper/handleLink';
import { resizeBrowser } from '@/presentation/helper/swiffy';
import VArticleHeader from '@/layout/header/VArticleHeader.vue';
import VPage from '@/layout/page/VPage.vue';
import VInputButton from '@/molecule/form/VInputButton/VInputButton.vue';
import VAccordion from '@/atom/accordion/VAccordion.vue';
import VButton from '@/atom/button/VButton/VButton.vue';
import VCheckbox from '@/atom/checkbox/VCheckbox/VCheckbox.vue';
import VDownloadContentWrapper from '@/atom/contentWrapper/VDownloadContentWrapper.vue';
import VIcon from '@/atom/icon/VIcon/VIcon.vue';
import VLoader from '@/atom/loader/VLoader/VLoader.vue';
import VModal from '@/atom/modal/VModal/VModal.vue';
import VRefresher from '@/atom/refresher/VRefresher/VRefresher.vue';
import VAnimationScript from '@/atom/script/VAnimationScript/VAnimationScript.vue';
import VBundle from '@/atom/script/VBundle/VBundle.vue';
import VH5p from '@/atom/script/VH5p/VH5p.vue';

interface Props {
    id: string;
}
interface DownloadFile {
    id: string;
    url: string;
    size: string;
}

const props = defineProps<Props>();
const isSwiffyReady: Ref<boolean> = ref(false);
const articleState = useArticleState();
const addToCollectionsModal = ref();
const createCollectionModal = ref();
const rawHtmlLinksContainer = ref();
const tempCollectionsToUpdate: Collection[] = [];
const { data: pageData, getData: getDownloadTopicBranch } = useDownloadTopicBranch(props.id);

const { getData: getDownloadLive } = useDownloadLive({
    params: {
        id: props.id,
    },
});

function openCollectionModal(): void {
    articleState.getCollections();
    tempCollectionsToUpdate.length = 0;
    addToCollectionsModal.value?.$el.present();
}

function toggleArticleInTempList(event: CustomEvent, item: Collection): void {
    if (event.detail.checked === true) {
        tempCollectionsToUpdate.push(item);
    }

    if (event.detail.checked === false) {
        const filtered = tempCollectionsToUpdate.filter(
            (collection) => !(collection.id === item.id && collection.type === item.type),
        );

        tempCollectionsToUpdate.length = 0;
        tempCollectionsToUpdate.push(...filtered);
    }
}

function handleModalConfirmation(): void {
    if (!pageData.value || !pageData.value.download) {
        return;
    }

    articleState.saveArticlesToCollections(pageData.value.download, tempCollectionsToUpdate);
}

function handleCreateCollectionSubmit({ value }: { value: string }): void {
    if (!pageData.value || !pageData.value.download || !value) {
        return;
    }

    articleState.createAndAddArticleToNewCollection(pageData.value.download, value);
}

function openCreateCollectionModal(): void {
    // addToCollectionsModal.value?.$el.dismiss();
    createCollectionModal.value?.$el.present();
}

function setSwiffiyToReady() {
    isSwiffyReady.value = true;
}

function getFileSizeInBites(size: number): string {
    const sizes = ['B', 'KB', 'MB', 'GB'];
    const kb = 1024;
    let i = 0;

    do {
        size = size / kb;
        i++;
    } while (size >= kb);

    return `${size.toFixed(2)} ${sizes[i]}`;
}

const downloadedFiles = computed<DownloadFile[]>(() => {
    if (!pageData.value || !pageData.value.download) {
        return [];
    }

    return pageData.value.download.download_file.map((file) => ({
        id: file.id,
        size: `Größe: ${getFileSizeInBites(file.size)}`,
        url: file.url,
    }));
});

async function handleRefresh(event: RefresherCustomEvent): Promise<void> {
    await createRefresherHandler({
        event: event,
        callback: async () => {
            await getDownloadLive();
            await getDownloadTopicBranch();
        },
    });
}

watch(rawHtmlLinksContainer, () => {
    if (!rawHtmlLinksContainer.value) {
        return;
    }

    handleLinksClick(rawHtmlLinksContainer.value);
});

onIonViewWillEnter(async () => {
    addDownloadToHistory(props.id);

    await getDownloadTopicBranch();

    resizeBrowser();
});
</script>

<template>
    <v-page modifier="v-pt-0 md:v-pt-0">
        <template v-if="pageData">
            <v-bundle />

            <v-h5p />

            <v-animation-script :onload="setSwiffiyToReady" />
        </template>

        <v-loader :logs="['useDownload']" />

        <template #refresher>
            <v-refresher @ion-refresh="handleRefresh" />
        </template>

        <template #header>
            <v-article-header
                :id="id"
                :type="ContentType.DOWNLOAD"
                @on-open-add-to-collection="openCollectionModal"
            >
                <template v-if="pageData && pageData.topic && pageData.branch" #pre-title>
                    <router-link
                        :to="{ name: RouteName.BRANCH_ID, params: { id: pageData.branch.id } }"
                    >
                        {{ pageData.branch.name }}
                    </router-link>
                    /
                    <router-link
                        :to="{ name: RouteName.TOPIC_ID, params: { id: pageData.topic.id } }"
                    >
                        {{ pageData.topic.name }}
                    </router-link>
                </template>

                <template v-if="pageData && pageData.download">
                    {{ pageData.download.title }}
                </template>
            </v-article-header>
        </template>

        <template v-if="pageData">
            <h1 class="v-pt-box-xl v-text-h2">{{ $t.download.headline }}</h1>

            <p class="v-p v-mt-box-lg">
                <span class="v-font-bold">Typ:</span>

                {{ pageData.download.download_type }}
            </p>

            <div
                v-if="isSwiffyReady"
                class="site v-mt-box-xl v-border-t v-border-t-gray-80 v-pt-box-xl"
            >
                <div v-html="pageData.download.download_description" />

                <div
                    v-if="pageData.download.animation_code?.rendered_string"
                    v-html="pageData.download.animation_code.rendered_string"
                />

                <template v-if="pageData.download.animation_code?.attachments">
                    <component
                        :is="attachment.tag"
                        v-for="(attachment, indexAttachment) in pageData.download.animation_code
                            .attachments"
                        :key="`${indexAttachment}-task-question-attachment`"
                        :type="attachment.type"
                    >
                        {{ attachment.content }}
                    </component>
                </template>
            </div>

            <div class="site">
                <template v-if="downloadedFiles.length">
                    <template v-for="file in downloadedFiles" :key="file.id">
                        <p class="v-p v-mt-box-lg">{{ file.size }}</p>

                        <a v-if="file.url" :href="file.url" target="_blank">
                            <v-button class="v-w-full" size="small" :is-rounded="false">{{
                                $t.download.fielDownloadButtonLabel
                            }}</v-button>
                        </a>
                    </template>
                </template>
            </div>

            <v-accordion v-if="pageData.download.download_didactic_info.trim().length">
                <template #label>{{ $t.download.accordionLabel }}:</template>

                <v-download-content-wrapper>
                    <div v-html="pageData.download.download_didactic_info" />
                </v-download-content-wrapper>
            </v-accordion>
        </template>

        <v-modal ref="addToCollectionsModal" with-save @on-confirm="handleModalConfirmation">
            <template #title>{{ $t.collection.addCollection }}</template>

            <v-button modifier="v-w-full" @click="openCreateCollectionModal">
                <v-icon src="icon/add-collection.svg" with-space-right />

                {{ $t.collection.createCollection }}
            </v-button>

            <template v-if="articleState.collections && articleState.collections.length">
                <div
                    v-for="collection in articleState.collections"
                    :key="collection.id"
                    class="v-mt-box-md v-flex v-items-center v-rounded v-border v-border-gray-20 v-bg-white v-p-box"
                >
                    <v-checkbox
                        :disabled="
                            findArticleInCollection(collection, props.id, ContentType.DOWNLOAD)
                        "
                        :checked="
                            findArticleInCollection(collection, props.id, ContentType.DOWNLOAD)
                        "
                        modifier="v-mr-auto v-w-full v-h3 v-font-bold"
                        @ion-change="(event) => toggleArticleInTempList(event, collection)"
                    >
                        {{ collection.private_title }}
                    </v-checkbox>
                </div>
            </template>
        </v-modal>

        <v-modal ref="createCollectionModal">
            <template #title>{{ $t.collection.addCollection }}</template>

            <v-input-button
                icon="icon/add-collection.svg"
                :button-text="$t.collection.create"
                :placeholder="$t.collection.collectionName"
                @on-submit="handleCreateCollectionSubmit"
            />
        </v-modal>
    </v-page>
</template>
