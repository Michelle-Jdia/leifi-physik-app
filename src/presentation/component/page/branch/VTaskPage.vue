<script setup lang="ts">
import type { Collection } from '@/data/type/app/collection';
import type { Issue } from '@/data/type/app/issue';
import type { RefresherCustomEvent } from '@ionic/vue';
import type { Ref } from 'vue';
import { onIonViewWillEnter } from '@ionic/vue';
import { ref, watch } from 'vue';
import { $t } from '@/application/translation';
import { ContentType } from '@/data/type/helper/contentType';
import { createReactiveData } from '@/connection/helper/fetcher';
import { getContextualRouteNameByType } from '@/connection/helper/navigation';
import { createRefresherHandler } from '@/connection/helper/refresher';
import { useTaskTopicBranch } from '@/connection/helper/task';
import { RouteName } from '@/connection/router/routeName';
import { useArticleState } from '@/state/articleState';
import { findArticleInCollection } from '@/store/collectionStore';
import { addTaskToHistory } from '@/store/historyStore';
import { useIssue } from '@/store/issueStore';
import { useTaskLive } from '@/store/taskStore';
import { handleLinksClick } from '@/presentation/helper/handleLink';
import { resizeBrowser } from '@/presentation/helper/swiffy';
import VArticleHeader from '@/layout/header/VArticleHeader.vue';
import VPage from '@/layout/page/VPage.vue';
import VInputButton from '@/molecule/form/VInputButton/VInputButton.vue';
import VAccordion from '@/atom/accordion/VAccordion.vue';
import VButton from '@/atom/button/VButton/VButton.vue';
import VCheckbox from '@/atom/checkbox/VCheckbox/VCheckbox.vue';
import VIcon from '@/atom/icon/VIcon/VIcon.vue';
import VLinkItem from '@/atom/item/VLinkItem/VLinkItem.vue';
import VTextIconItem from '@/atom/item/VTextIconItem/VTextIconItem.vue';
import VLoader from '@/atom/loader/VLoader/VLoader.vue';
import VModal from '@/atom/modal/VModal/VModal.vue';
import VRefresher from '@/atom/refresher/VRefresher/VRefresher.vue';
import VAnimationScript from '@/atom/script/VAnimationScript/VAnimationScript.vue';
import VBundle from '@/atom/script/VBundle/VBundle.vue';
import VH5p from '@/atom/script/VH5p/VH5p.vue';

interface Props {
    id: string;
}

const props = defineProps<Props>();
const articleState = useArticleState();
const addToCollectionsModal = ref();
const createCollectionModal = ref();
const tempCollectionsToUpdate: Collection[] = [];
const isSwiffyReady = ref(false);
const refAnchors: Ref<HTMLHeadingElement[]> = ref([]);
const isPageDataReady = ref(false);
const rawHtmlLinksContainer = ref();
const { data: pageData, getData: getTaskTopicBranch } = useTaskTopicBranch(props.id);

const { getData: getTaskLive } = useTaskLive({
    params: {
        id: props.id,
    },
});

const { data: issueData, getData: getIssueData } = createReactiveData<Issue | void>(async () => {
    await getTaskTopicBranch();

    isPageDataReady.value = true;

    if (
        !pageData.value ||
        !pageData.value.task.referenced_issue.length ||
        !pageData.value.task.referenced_issue[0]
    ) {
        return;
    }

    const { data: issue, getData: getIssue } = useIssue({
        params: {
            id: pageData.value.task.referenced_issue[0], // @TODO: array of related issues?
        },
    });

    await getIssue();

    if (!issue.value) {
        return;
    }

    return issue.value;
})();

function openCollectionModal(): void {
    articleState.getCollections();
    tempCollectionsToUpdate.length = 0;
    addToCollectionsModal.value?.$el.present();
}

function toggleArticleInTempList(event: CustomEvent, item: Collection): void {
    if (event.detail.checked === true) {
        tempCollectionsToUpdate.push(item);
    }

    if (event.detail.checked === false) {
        const filtered = tempCollectionsToUpdate.filter(
            (collection) => !(collection.id === item.id && collection.type === item.type),
        );

        tempCollectionsToUpdate.length = 0;
        tempCollectionsToUpdate.push(...filtered);
    }
}

function openCreateCollectionModal(): void {
    // addToCollectionsModal.value?.$el.dismiss();
    createCollectionModal.value?.$el.present();
}

function handleModalConfirmation(): void {
    if (!pageData.value || !pageData.value.task) {
        return;
    }

    articleState.saveArticlesToCollections(pageData.value.task, tempCollectionsToUpdate);
}

function handleCreateCollectionSubmit({ value }: { value: string }): void {
    if (!value || !pageData.value || !pageData.value.task) {
        return;
    }

    articleState.createAndAddArticleToNewCollection(pageData.value.task, value);
}

const addRefAnchor = (index: number) => (anchor: any) => {
    // we need to use index here, otherwise the loops could spam multiple refs in the array
    refAnchors.value[index] = anchor;
};

function setSwiffiyToReady() {
    isSwiffyReady.value = true;
}

async function handleRefresh(event: RefresherCustomEvent): Promise<void> {
    await createRefresherHandler({
        event: event,
        callback: async () => {
            await getTaskLive();
            await getIssueData();
        },
    });
}

addTaskToHistory(props.id);

onIonViewWillEnter(async () => {
    await getIssueData();

    resizeBrowser();
});

watch(isPageDataReady, () => {
    if (!isPageDataReady.value) {
        return;
    }

    handleLinksClick(rawHtmlLinksContainer.value);
});
</script>

<template>
    <v-page modifier="v-pt-0 md:v-pt-0">
        <template v-if="isPageDataReady">
            <v-bundle />

            <v-h5p />

            <v-animation-script :onload="setSwiffiyToReady" />
        </template>

        <v-loader :logs="['useTask']" />

        <template #refresher>
            <v-refresher @ion-refresh="handleRefresh" />
        </template>

        <template #header>
            <v-article-header
                :id="id"
                :anchors="refAnchors"
                :type="ContentType.TASK"
                @on-open-add-to-collection="openCollectionModal"
            >
                <template v-if="pageData && pageData.topic && pageData.branch" #pre-title>
                    <router-link
                        :to="{ name: RouteName.BRANCH_ID, params: { id: pageData.branch.id } }"
                    >
                        {{ pageData.branch.name }}
                    </router-link>
                    /
                    <router-link
                        :to="{ name: RouteName.TOPIC_ID, params: { id: pageData.topic.id } }"
                    >
                        {{ pageData.topic.name }}
                    </router-link>
                </template>

                {{ pageData?.task?.title }}
            </v-article-header>
        </template>

        <template v-if="pageData?.task">
            <h1 :ref="addRefAnchor(0)" class="v-pt-box-xl v-text-h2">{{ $t.task.headline }}</h1>

            <v-text-icon-item modifier="v-mt-box-lg" :level="pageData.task.level.color">
                <span class="v-text-copy v-font-bold">Schwierigkeitsgrad: </span>

                <span class="v-text-copy">
                    {{ pageData.task.level.name }}
                </span>
            </v-text-icon-item>

            <div class="v-p v-mt-2">
                <span class="v-text-copy v-font-bold"> Typ: </span>

                <span class="v-text-copy"> {{ pageData.task.task_type }} </span>
            </div>

            <div
                v-if="isSwiffyReady && pageData.task.question.length"
                class="site v-mt-box-xl v-border-t v-border-t-gray-80 v-pt-box-xl"
            >
                <div
                    v-for="(question, index) in pageData.task.question"
                    :key="`${index}-task-question`"
                >
                    <div ref="rawHtmlLinksContainer" v-html="question.rendered_string" />

                    <component
                        :is="attachment.tag"
                        v-for="(attachment, indexAttachment) in question.attachments"
                        :key="`${indexAttachment}-task-question-attachment`"
                        :type="attachment.type"
                    >
                        {{ attachment.content }}
                    </component>
                </div>
            </div>

            <div
                v-if="pageData.task.paragraphs"
                ref="rawHtmlLinksContainer"
                class="site v-mt-box"
                v-html="pageData.task.paragraphs"
            />

            <template v-if="pageData.task.answer.length">
                <h3 class="v-h3 v-mt-section v-text-primary">
                    {{ $t.task.solutionSection.headline }}
                </h3>

                <v-accordion>
                    <template #label>{{ $t.task.solutionSection.accordionLabel }}</template>

                    <div
                        v-for="(answer, index) in pageData.task.answer"
                        :key="`${index}-task-answer`"
                        ref="rawHtmlLinksContainer"
                        class="site"
                        v-html="answer"
                    />
                </v-accordion>
            </template>

            <template v-if="issueData">
                <h3 :ref="addRefAnchor(1)" class="v-h3 v-pt-box-xl v-text-primary">
                    {{ $t.task.relatedIssueTitle }}
                </h3>

                <router-link
                    class="v-mt-box v-block"
                    :to="{
                        name: getContextualRouteNameByType(issueData.type),
                        params: { id: issueData.id },
                    }"
                >
                    <v-link-item>{{ issueData.title }}</v-link-item>
                </router-link>
            </template>
        </template>

        <v-modal ref="addToCollectionsModal" with-save @on-confirm="handleModalConfirmation">
            <template #title>{{ $t.collection.addCollection }}</template>

            <v-button modifier="v-w-full" @click="openCreateCollectionModal">
                <v-icon src="icon/add-collection.svg" with-space-right />

                {{ $t.collection.createCollection }}
            </v-button>

            <template v-if="articleState.collections && articleState.collections.length">
                <div
                    v-for="collection in articleState.collections"
                    :key="collection.id"
                    class="v-mt-box-md v-flex v-items-center v-rounded v-border v-border-gray-20 v-bg-white v-p-box"
                >
                    <v-checkbox
                        :disabled="findArticleInCollection(collection, props.id, ContentType.TASK)"
                        :checked="findArticleInCollection(collection, props.id, ContentType.TASK)"
                        modifier="v-mr-auto v-w-full v-h3 v-font-bold"
                        @ion-change="(event) => toggleArticleInTempList(event, collection)"
                    >
                        {{ collection.private_title }}
                    </v-checkbox>
                </div>
            </template>
        </v-modal>

        <v-modal ref="createCollectionModal">
            <template #title>{{ $t.collection.addCollection }}</template>

            <v-input-button
                icon="icon/add-collection.svg"
                :button-text="$t.collection.create"
                :placeholder="$t.collection.collectionName"
                @on-submit="handleCreateCollectionSubmit"
            />
        </v-modal>
    </v-page>
</template>
