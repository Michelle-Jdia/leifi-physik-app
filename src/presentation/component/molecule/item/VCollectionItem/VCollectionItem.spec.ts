import { render } from '@testing-library/vue';
import { describe, expect, test } from 'vitest';
import { testSnapshot } from '@/presentation/helper/test';
import { topic } from '@/presentation/static/topic';
import { branch } from '@/presentation/static/branch';
import VCollectionItem from '@/molecule/item/VCollectionItem/VCollectionItem.vue';

describe('VCollectionItem', () => {
    const options = {
        props: {
            title: topic.title,
            branch: {
                title: branch.title,
                color: '',
            },
        },
        global: {
            stubs: {
                ionIcon: true,
            },
        },
    };

    test('shows the correct text', () => {
        const { getByText, queryByTestId } = render(VCollectionItem, options);

        getByText(topic.title);
        getByText(branch.title);

        expect(queryByTestId('collection-item-notification')).toBeFalsy();
        expect(queryByTestId('collection-item-feed')).toBeFalsy();
    });

    test('doesnt show the text if it was not passed', () => {
        const { queryByText, getByTestId } = render(VCollectionItem, {
            ...options,
            props: {
                title: '',
                withNotification: true,
            },
        });

        expect(queryByText(topic.title)).toBeFalsy();
        expect(queryByText(branch.title)).toBeFalsy();

        getByTestId('collection-item-notification');
        getByTestId('collection-item-feed');
    });

    testSnapshot(VCollectionItem, options);
});
