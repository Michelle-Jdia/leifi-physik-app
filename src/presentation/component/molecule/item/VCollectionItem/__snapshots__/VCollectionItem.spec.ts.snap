// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`VCollectionItem > snapshot 1`] = `
<div
  class="collection-item v-relative v-flex v-items-start v-justify-between v-rounded-sm v-border v-border-gray-20 v-bg-white v-p-box v-shadow"
>
  <div>
    <h3
      class="v-h3"
    >
      Gleichf√∂rmige Bewegungen
    </h3>
    <p
      class="v-p v-mt-1 v-text-gray"
    >
      2qe485
    </p>
  </div>
  <!--v-if-->
  <!--v-if-->
  <!--v-if-->
</div>
`;
